File: src/Java/20.ValidParentheses.java
import java.awt.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Stack;

public class ValidParentheses {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        int n = s.length();
        if(n % 2 == 1){
            return false;
        }
        for(int i = 0;i < n; ++i){
            Character c = s.charAt(i);
            if(c == '(' || c == '[' || c == '{'){
                stack.push(c);
            }
            else{
                if(c == ')' || c == ']' || c == '}'){
                    if(stack.empty() == true)
                        return false;
                    Character last = stack.pop();
                    if(last == '(' && c == ')'){
                        continue;
                    }
                    if(last == '[' && c == ']'){
                        continue;
                    }
                    if(last == '{' && c == '}'){
                        continue;
                    }
                    else{
                        return false;
                    }
                }
                else{
                    return false;
                }
            }
        }
        return stack.empty();
    }
}

File: src/CSharp/1-100/20.ValidParentheses.cs

using System;
using System.Linq;
using System.Collections;
using System.Collections.Generic;

namespace PeefyLeetCode.ValidParentheses
{
    public class Solution
    {
        public bool IsValid(string s)
        {
            var n = s.Length;
            // 开辟一个数组充当stack
            char[] a = new char[n / 2 + 1];
            var top = 0;
            for (var i = 0; i < n; ++i)
            {
                var c = s[i];
                switch (c)
                {
                    case '(':
                    case '{':
                    case '[':
                        a[top++] = c;
                        if (top >= n / 2 + 1)
                            return false;
                        break;
                    case ')':
                        if (top > 0)
                            if (a[top - 1] == '(')
                            {
                                top--;
                                break;
                            }
                        return false;
                    case '}':
                        if (top > 0)
                            if (a[top - 1] == '{')
                            {
                                top--;
                                break;
                            }
                        return false;
                    case ']':
                        if (top > 0)
                            if (a[top - 1] == '[')
                            {
                                top--;
                                break;
                            }
                        return false;

                    default:
                        return false;
                }
            }
            return !(top > 0);
        }

    }
}
